File: src/backend/ClarityDQ.Core/Entities/Rule.cs
namespace ClarityDQ.Core.Entities;

public class Rule
{
    public Guid Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;
    public RuleType Type { get; set; }
    public string WorkspaceId { get; set; } = string.Empty;
    public string DatasetName { get; set; } = string.Empty;
    public string TableName { get; set; } = string.Empty;
    public string ColumnName { get; set; } = string.Empty;
    public string Expression { get; set; } = string.Empty;
    public double Threshold { get; set; }
    public RuleSeverity Severity { get; set; }
    public bool IsEnabled { get; set; }
    public DateTime CreatedAt { get; set; }
    public DateTime? UpdatedAt { get; set; }
    public string CreatedBy { get; set; } = string.Empty;
}

public enum RuleType
{
    Completeness,
    Accuracy,
    Consistency,
    Uniqueness,
    Validity,
    Custom
}

public enum RuleSeverity
{
    Low,
    Medium,
    High,
    Critical
}

public class RuleExecution
{
    public Guid Id { get; set; }
    public Guid RuleId { get; set; }
    public Rule? Rule { get; set; }
    public DateTime ExecutedAt { get; set; }
    public RuleExecutionStatus Status { get; set; }
    public long RecordsChecked { get; set; }
    public long RecordsPassed { get; set; }
    public long RecordsFailed { get; set; }
    public double SuccessRate { get; set; }
    public string? ResultDetails { get; set; }
    public string? ErrorMessage { get; set; }
    public int DurationMs { get; set; }
}

public enum RuleExecutionStatus
{
    Pending,
    Running,
    Completed,
    Failed,
    Skipped
}

File: src/frontend/src/types/rule.ts
export enum RuleType {
  Completeness = 0,
  Accuracy = 1,
  Consistency = 2,
  Uniqueness = 3,
  Validity = 4,
  Custom = 5,
}

export enum RuleSeverity {
  Low = 0,
  Medium = 1,
  High = 2,
  Critical = 3,
}

export enum RuleExecutionStatus {
  Pending = 0,
  Running = 1,
  Completed = 2,
  Failed = 3,
  Skipped = 4,
}

export interface Rule {
  id: string;
  name: string;
  description: string;
  type: RuleType;
  workspaceId: string;
  datasetName: string;
  tableName: string;
  columnName?: string;
  expression: string;
  threshold: number;
  severity: RuleSeverity;
  isEnabled: boolean;
  createdAt: string;
  updatedAt?: string;
  createdBy: string;
}

export interface RuleExecution {
  id: string;
  ruleId: string;
  executedAt: string;
  status: RuleExecutionStatus;
  recordsChecked: number;
  recordsPassed: number;
  recordsFailed: number;
  successRate: number;
  resultDetails?: string;
  errorMessage?: string;
  durationMs: number;
}

export interface CreateRuleRequest {
  name: string;
  description: string;
  type: RuleType;
  workspaceId: string;
  datasetName: string;
  tableName: string;
  columnName?: string;
  expression: string;
  threshold: number;
  severity: RuleSeverity;
  isEnabled: boolean;
}

export interface UpdateRuleRequest {
  name: string;
  description: string;
  expression: string;
  threshold: number;
  severity: RuleSeverity;
  isEnabled: boolean;
}
